File: src/analyzer/ComplexityAnalyzer.Core/ComposedCost.cs
namespace ComplexityAnalyzer.Core;

/// <summary>
/// Intermediate cost of a statement or subtree. Sequential composition
/// adds time and takes peak space; conditionals take the worst branch.
/// </summary>
public sealed record ComposedCost
{
    public required ComplexityExpression Time { get; init; }

    public required ComplexityExpression Space { get; init; }

    public AnalysisConfidence Confidence { get; init; } =
        AnalysisConfidence.High;

    public required ComplexityEvidence Evidence { get; init; }

    public IReadOnlyList<AnalysisWarning> Warnings { get; init; } = [];

    public IReadOnlyList<BoundingSuggestion> Suggestions { get; init; } = [];

    public static ComposedCost Unit(string kind, string label, LineSpan? span)
    {
        return new ComposedCost
        {
            Time = Cx.One,
            Space = Cx.One,
            Evidence = ComplexityEvidence.Leaf(kind, label, Cx.One, span),
        };
    }

    public static ComposedCost Of(
        ComplexityExpression time,
        ComplexityExpression space,
        string kind,
        string label,
        LineSpan? span,
        AnalysisConfidence confidence = AnalysisConfidence.High)
    {
        return new ComposedCost
        {
            Time = time,
            Space = space,
            Confidence = confidence,
            Evidence = ComplexityEvidence.Leaf(kind, label, time, span),
        };
    }

    public static AnalysisConfidence Min(
        AnalysisConfidence a, AnalysisConfidence b) =>
        a < b ? a : b;
}

File: src/extension/src/analysis/typescript/engine/composedCost.ts
import { One } from './cx';
import type { ComplexityExpression } from './expression';
import {
  evidenceLeaf,
  minConfidence,
  type AnalysisConfidence,
  type AnalysisWarning,
  type BoundingSuggestion,
  type ComplexityEvidence,
  type LineSpan,
} from './types';

export interface ComposedCost {
  time: ComplexityExpression;
  space: ComplexityExpression;
  confidence: AnalysisConfidence;
  evidence: ComplexityEvidence;
  warnings: readonly AnalysisWarning[];
  suggestions: readonly BoundingSuggestion[];
}

export function unitCost(
  kind: string,
  label: string,
  span?: LineSpan,
): ComposedCost {
  return {
    time: One,
    space: One,
    confidence: 'high',
    evidence: evidenceLeaf(kind, label, One, span),
    warnings: [],
    suggestions: [],
  };
}

export function ofCost(
  time: ComplexityExpression,
  space: ComplexityExpression,
  kind: string,
  label: string,
  span?: LineSpan,
  confidence: AnalysisConfidence = 'high',
): ComposedCost {
  return {
    time,
    space,
    confidence,
    evidence: evidenceLeaf(kind, label, time, span),
    warnings: [],
    suggestions: [],
  };
}

export { minConfidence };
